### DISCUSSION TITLE
Key for different node and link styles to enhance UX of your graph

### DISCUSSION THREAD

**Discussion by ndtreviv** [f5/unovis#110] [Ideas]
I'm creating a graph (force graph to be exact) showing the relationship between different types of entities that I'm pulling out of a graph database.

Each entity type has a different shape. Nodes within one entity type are sized and coloured differently according to certain properties (read: more "interesting" nodes are bigger and more colourful).

I want to format links differently to distinguish between different types of graph edge (ownedBy, affiliatedWith, resolvesTo etc).

As I'm doing this I'm thinking: I need a key to show the user what the different link styles actually _mean_.

It would be great if there was a `<VisKey ... />` object that used all the same nodeFill etc functions to generate a key-based visualisation of each different node and link type/style.

OR, failing that, some way to render a single representation of a node or link rather than an entire graph...

ATM I'm thinking of creating a key that has a collection of `<VisGraph.../>` objects with hammed in data to show the different node styles.

**Comment by rokotyan** [comment]
Nice idea, we'll definitely consider it when we'll be working on https://github.com/f5/unovis/issues/24

**Reply by ndtreviv** [reply]
So ATM I've got two `<VisGraph... />`'s showing a key. One shows the different node type formats in a parallel configuration. The other shows the different link formats. Both have dragging and zooming turned off so they're effectively read-only/immutable by the user.

I'm not sure now how you'd automate this without knowing all the different permutations of return values to those functions that define node size/shape/colour etc. 

Doing what I'm doing now works quite well, except for the fact that the node labels are ellipsed (I'm using them to label different node formats with their types). If I could turn ellipsing off just for this use case that would be enough for me.

Also, the parallel view has properties that change the spacing between groups, but not between nodes within the same group. I'd like to make my key display more dense, but can't quite do that at the moment.

**Reply by rokotyan** [reply]
@ndtreviv That sounds great, can you post some screenshots? Just curious how it looks.

You can disable label trimming in version `1.0.3-beta.1`. See this [PR](https://github.com/f5/unovis/pull/121) for more details.

I'm planning to make the node spacing in parallel layouts configurable too, but a bit later.

**Reply by ndtreviv** [reply]
Sure, here's an example. Labels redacted for certain reasons.

<img width="348" alt="Screenshot 2023-01-26 at 21 53 54" src="https://user-images.githubusercontent.com/1530653/214959227-16515869-3a13-4f93-966f-4694cac258c5.png">

**Reply by rokotyan** [reply]
@ndtreviv Thanks!

**Reply by ndtreviv** [reply]
Looking at this reminds me - is there a way to move the arrowhead on the link closer towards the end of the link?

**Reply by rokotyan** [reply]
@ndtreviv Not as of now, but it would be a nice feature to have. I'll see how difficult it will be to implement.

**Reply by ndtreviv** [reply]
@rokotyan I've discovered a bug with doing the key this way. The links don't render correctly on the main graph unless the key is visible. The arrows are missing, and the link labels only show the first and last character when the key is hidden. Even when the key is revealed the link label doesn't render correctly. I've included some screen shots.

Link on main graph when the Key graph is hidden:
<img width="349" alt="Screenshot 2023-02-01 at 11 29 37" src="https://user-images.githubusercontent.com/1530653/216030886-3d6b972e-e976-46b2-838c-a9d2e51d3777.png">

Same link when the key graph is showing (we get arrows now, but text is wrong):
<img width="336" alt="Screenshot 2023-02-01 at 11 29 45" src="https://user-images.githubusercontent.com/1530653/216030984-f3a7a5a8-b587-492d-938c-c0e378b1a12d.png">

What it should look like (and looks like in the key):
<img width="291" alt="Screenshot 2023-02-01 at 11 29 49" src="https://user-images.githubusercontent.com/1530653/216031078-1f580d8d-5fb1-487d-8431-46540637b764.png">

**Reply by rokotyan** [reply]
@ndtreviv Could you please share an example reproducing that behavior over StackBlitz? I'm happy to take a look but from the screenshots it's not obvious what's going on.

**Reply by ndtreviv** [reply]
> @ndtreviv Could you please share an example reproducing that behavior over StackBlitz? I'm happy to take a look but from the screenshots it's not obvious what's going on.

Sure, I tried to recreate the exact issue but had even more issues along the way: https://stackblitz.com/edit/react-qt6n9k?file=src/App.js



If you uncomment the `linkLabel` property on line 240 you'll see all the links just disappear.
However, if you leave it commented out, notice how the arrows disappear whenever you expand the graph key accordion panel.

**Reply by rokotyan** [reply]
@ndtreviv Thanks, I'm planning to look into it next week.

**Reply by rokotyan** [reply]
@ndtreviv Fixed in `1.1.0-beta.2`
